File: server/src/main/java/com/hospitaltender/server/service/MarketResearchOfferService.java
package com.hospitaltender.server.service;

import com.hospitaltender.server.dto.request.CreateMarketResearchOfferRequest;
import com.hospitaltender.server.dto.request.UpdateMarketResearchOfferRequest;
import com.hospitaltender.server.dto.response.MarketResearchOfferResponse;

import java.math.BigDecimal;
import java.util.List;

public interface MarketResearchOfferService {

    MarketResearchOfferResponse create(CreateMarketResearchOfferRequest request);

    MarketResearchOfferResponse getById(Long id);

    List<MarketResearchOfferResponse> getByTenderItemId(Long tenderItemId);

    MarketResearchOfferResponse update(Long id, UpdateMarketResearchOfferRequest request);

    void delete(Long id);

    /**
     * Calculates the average price of all offers for a tender item
     * and updates the tender item's approximate unit cost.
     */
    BigDecimal calculateAndUpdateApproximateCost(Long tenderItemId);
}

File: ui/hospital-system-ui/src/services/market-research-offer.service.ts
import api from './api'
import {
  MarketResearchOfferResponse,
  CreateMarketResearchOfferRequest,
  UpdateMarketResearchOfferRequest,
} from '@/types'

export const marketResearchOfferService = {
  getByTenderItemId: (tenderItemId: number) =>
    api.get<MarketResearchOfferResponse[]>(`/market-research-offers/tender-item/${tenderItemId}`).then((res) => res.data),

  getById: (id: number) =>
    api.get<MarketResearchOfferResponse>(`/market-research-offers/${id}`).then((res) => res.data),

  create: (data: CreateMarketResearchOfferRequest) =>
    api.post<MarketResearchOfferResponse>('/market-research-offers', data).then((res) => res.data),

  update: (id: number, data: UpdateMarketResearchOfferRequest) =>
    api.put<MarketResearchOfferResponse>(`/market-research-offers/${id}`, data).then((res) => res.data),

  delete: (id: number) =>
    api.delete(`/market-research-offers/${id}`),

  calculateCost: (tenderItemId: number) =>
    api.post<number>(`/market-research-offers/tender-item/${tenderItemId}/calculate-cost`).then((res) => res.data),
}
